msmbuilder/lumping/bace.py

```diff
@@ -0,0 +1,380 @@
+                statesKeep, minX, minY, unmerged)
+
+            nCurrentStates -= 1
+
+            if self.save_all_maps:
+                saved_map = copy.deepcopy(macro_map)
+                self.map_dict[nCurrentStates] = saved_map
+
+            if nCurrentStates - 1 == self.n_macrostates:
+                self.microstate_mapping_ = macro_map
+
+    def partial_transform(self, sequence, mode='clip'):
+        if self.n_macrostates is None:
+            raise RuntimeError('n_macrostates must not be None to transform')
+        trimmed_sequence = super(BACE, self).partial_transform(sequence, mode)
+        if mode == 'clip':
+            return [self.microstate_mapping_[seq] for seq in trimmed_sequence]
+        elif mode == 'fill':
+            def nan_get(x):
+                try:
+                    x = int(x)
+                    return self.microstate_mapping_[x]
+                except ValueError:
+                    return np.nan
+
+            return np.asarray([nan_get(x) for x in trimmed_sequence])
+        else:
+            raise ValueError
+
+    @classmethod
+    def from_msm(cls, msm, n_macrostates, filter=1.1, save_all_maps=True,
+                 n_proc=1, chunk_size=100):
+        """Create and fit lumped model from pre-existing MSM.
+
+        Parameters
+        ----------
+        msm : MarkovStateModel
+            The input microstate msm to use.
+        n_macrostates : int
+            The number of macrostates
+
+        Returns
+        -------
+        lumper : cls
+            The fit MVCA object.
+        """
+        params = msm.get_params()
+        lumper = cls(n_macrostates, filter, save_all_maps, n_proc,
+                     chunk_size, **params)
+
+        lumper.transmat_ = msm.transmat_
+        lumper.populations_ = msm.populations_
+        lumper.mapping_ = msm.mapping_
+        lumper.countsmat_ = msm.countsmat_
+        lumper.n_states_ = msm.n_states_
+
+        if n_macrostates is not None:
+            lumper._do_lumping()
+
+        return lumper
+
+    def getInds(self, c, stateInds, updateSingleState=None):
+        indices = []
+        for s in stateInds:
+            dest = np.where(c[s, :] > 1)[0]
+
+            if updateSingleState != None:
+                dest = dest[np.where(dest != updateSingleState)[0]]
+            else:
+                dest = dest[np.where(dest > s)[0]]
+
+            if dest.shape[0] == 0:
+                continue
+            elif dest.shape[0] < self.chunk_size:
+                indices.append((s, dest))
+            else:
+                i = 0
+                while dest.shape[0] > i:
+                    if i + self.chunk_size > dest.shape[0]:
+                        indices.append((s, dest[i:]))
+                    else:
+                        indices.append((s, dest[i:i + self.chunk_size]))
+                    i += self.chunk_size
+        return indices
+
+    def mergeTwoClosestStates(self, c, w, indRecalc, dMat,
```

Comment:
Should class functions such as this one be hidden from users (e.g. `_mergeTwoClosestStates`)?

Reply:
yeah that's a good idea I'll hide them all

Reply:
👍 LGTM otherwise